// react-stylesheet/lib/attachStylesheet.js
/**
 * @copyright 2015 Prometheus Research, LLC
 */

'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

exports['default'] = attachStylesheet;

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj['default'] = obj; return newObj; } }

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _invariant = require('invariant');

var _invariant2 = _interopRequireDefault(_invariant);

var _getComponentDisplayName = require('./getComponentDisplayName');

var _getComponentDisplayName2 = _interopRequireDefault(_getComponentDisplayName);

var _isValidReactComponent = require('./isValidReactComponent');

var _isValidReactComponent2 = _interopRequireDefault(_isValidReactComponent);

var _transferStaticProperties = require('./transferStaticProperties');

var _transferStaticProperties2 = _interopRequireDefault(_transferStaticProperties);

var _Stylesheet = require('./Stylesheet');

var Stylesheet = _interopRequireWildcard(_Stylesheet);

function attachStylesheet(Component, spec) {
  if ((0, _isValidReactComponent2['default'])(Component)) {
    return attachStylesheetImpl(Component, spec);
  } else if (spec === undefined) {
    var _ret = (function () {
      var spec = Component;
      return {
        v: function attachStylesheetDecorator(Component) {
          return attachStylesheetImpl(Component, spec);
        }
      };
    })();

    if (typeof _ret === 'object') return _ret.v;
  } else {
    (0, _invariant2['default'])(false, 'Styleable(...): can only be used as a decorator or factory function');
  }
}

function attachStylesheetImpl(Component, spec) {
  var StyleableComponent = (function (_StyleableComponentDecorator) {
    _inherits(StyleableComponent, _StyleableComponentDecorator);

    function StyleableComponent() {
      _classCallCheck(this, StyleableComponent);

      _get(Object.getPrototypeOf(StyleableComponent.prototype), 'constructor', this).apply(this, arguments);
    }

    _createClass(StyleableComponent, null, [{
      key: 'displayName',
      value: 'Styleable(' + (0, _getComponentDisplayName2['default'])(Component) + ')',
      enumerable: true
    }, {
      key: 'stylesheet',
      value: Stylesheet.createStylesheet(spec),
      enumerable: true
    }, {
      key: 'Component',
      value: Component,
      enumerable: true
    }]);

    return StyleableComponent;
  })(StyleableComponentDecorator);
  (0, _transferStaticProperties2['default'])(Component, StyleableComponent, ['Component', 'stylesheet']);
  return StyleableComponent;
}

var StyleableComponentDecorator = (function (_React$Component) {
  _inherits(StyleableComponentDecorator, _React$Component);

  function StyleableComponentDecorator() {
    _classCallCheck(this, StyleableComponentDecorator);

    _get(Object.getPrototypeOf(StyleableComponentDecorator.prototype), 'constructor', this).apply(this, arguments);
  }

  _createClass(StyleableComponentDecorator, [{
    key: 'render',
    value: function render() {
      var _constructor = this.constructor;
      var Component = _constructor.Component;
      var stylesheet = _constructor.stylesheet;

      return _react2['default'].createElement(Component, _extends({ stylesheet: stylesheet }, this.props));
    }
  }], [{
    key: 'style',
    value: function style(spec) {
      var _this = this;

      var stylesheet = Stylesheet.overrideStylesheet(this.stylesheet, spec);
      var StyleableComponent = (function (_StyleableComponentDecorator2) {
        _inherits(StyleableComponent, _StyleableComponentDecorator2);

        function StyleableComponent() {
          _classCallCheck(this, StyleableComponent);

          _get(Object.getPrototypeOf(StyleableComponent.prototype), 'constructor', this).apply(this, arguments);
        }

        _createClass(StyleableComponent, null, [{
          key: 'stylesheet',
          value: stylesheet,
          enumerable: true
        }, {
          key: 'Component',
          value: _this.Component,
          enumerable: true
        }, {
          key: 'displayName',
          value: _this.displayName,
          enumerable: true
        }]);

        return StyleableComponent;
      })(StyleableComponentDecorator);
      (0, _transferStaticProperties2['default'])(this.Component, StyleableComponent, ['Component', 'stylesheet']);
      return StyleableComponent;
    }
  }, {
    key: 'stylesheet',
    value: null,
    enumerable: true
  }, {
    key: 'Component',
    value: null,
    enumerable: true
  }]);

  return StyleableComponentDecorator;
})(_react2['default'].Component);

module.exports = exports['default'];
