// react-stylesheet/lib/transferStaticProperties.js
/**
 * @copyright 2015, Prometheus Research, LLC
 */

// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = transferStaticProperties;
var FUNCTION_STATIC_PROPERTIES = {
  arguments: true,
  arity: true,
  caller: true,
  length: true,
  name: true,
  displayName: true,
  prototype: true
};

/**
 * Transfer all static properties and methods from `source` to `target`.
 */

function transferStaticProperties(source, target) {
  var ignoreKeys = arguments.length <= 2 || arguments[2] === undefined ? null : arguments[2];

  var keys = Object.getOwnPropertyNames(source);
  for (var i = 0; i < keys.length; i++) {
    var key = keys[i];
    if (!source.hasOwnProperty(key)) {
      continue;
    }
    if (FUNCTION_STATIC_PROPERTIES[key]) {
      continue;
    }
    if (ignoreKeys && ignoreKeys.indexOf(key) > -1) {
      continue;
    }
    target[key] = source[key];
  }
}

module.exports = exports["default"];
