// react-stylesheet/lib/Utils.js
/**
 * @copyright 2015, Prometheus Research, LLC
 */

'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _isArray2 = require('is-array');

var _isArray3 = _interopRequireDefault(_isArray2);

exports.isArray = _isArray3['default'];

var _isPlainObject2 = require('is-plain-object');

var _isPlainObject3 = _interopRequireDefault(_isPlainObject2);

exports.isPlainObject = _isPlainObject3['default'];

var _isFunction2 = require('is-function');

var _isFunction3 = _interopRequireDefault(_isFunction2);

exports.isFunction = _isFunction3['default'];

var _isString2 = require('is-string');

var _isString3 = _interopRequireDefault(_isString2);

exports.isString = _isString3['default'];

var _toSlugCase = require('to-slug-case');

var _toSlugCase2 = _interopRequireDefault(_toSlugCase);

exports.toDashCase = _toSlugCase2['default'];

var _ampUniqueId = require('amp-unique-id');

var _ampUniqueId2 = _interopRequireDefault(_ampUniqueId);

exports.uniqueID = _ampUniqueId2['default'];
